pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,1024 @@
+                            'IntervalArray.')
+        if limit is not None:
+            raise TypeError('limit is not supported for IntervalArray.')
+
+        if not isinstance(value, ABCInterval):
+            msg = ("'IntervalArray.fillna' only supports filling with a "
+                   "'scalar pandas.Interval'. Got a '{}' instead."
+                   .format(type(value).__name__))
+            raise TypeError(msg)
+
+        value = getattr(value, '_values', value)
+        self._check_closed_matches(value, name="value")
+
+        left = self.left.fillna(value=value.left)
+        right = self.right.fillna(value=value.right)
+        return self._shallow_copy(left, right)
+
+    @property
+    def dtype(self):
+        return IntervalDtype(self.left.dtype)
+
+    def astype(self, dtype, copy=True):
+        """
+        Cast to an ExtensionArray or NumPy array with dtype 'dtype'.
+
+        Parameters
+        ----------
+        dtype : str or dtype
+            Typecode or data-type to which the array is cast.
+
+        copy : bool, default True
+            Whether to copy the data, even if not necessary. If False,
+            a copy is made only if the old dtype does not match the
+            new dtype.
+
+        Returns
+        -------
+        array : ExtensionArray or ndarray
+            ExtensionArray or NumPy ndarray with 'dtype' for its dtype.
+        """
+        dtype = pandas_dtype(dtype)
+        if is_interval_dtype(dtype):
+            if dtype == self.dtype:
+                return self.copy() if copy else self
+
+            # need to cast to different subtype
+            try:
+                new_left = self.left.astype(dtype.subtype)
+                new_right = self.right.astype(dtype.subtype)
+            except TypeError:
+                msg = ('Cannot convert {dtype} to {new_dtype}; subtypes are '
+                       'incompatible')
+                raise TypeError(msg.format(dtype=self.dtype, new_dtype=dtype))
+            return self._shallow_copy(new_left, new_right)
+        elif is_categorical_dtype(dtype):
+            return Categorical(np.asarray(self))
+        # TODO: This try/except will be repeated.
+        try:
+            return np.asarray(self).astype(dtype, copy=copy)
+        except (TypeError, ValueError):
+            msg = 'Cannot cast {name} to dtype {dtype}'
+            raise TypeError(msg.format(name=type(self).__name__, dtype=dtype))
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        """
+        Concatenate multiple IntervalArray
+
+        Parameters
+        ----------
+        to_concat : sequence of IntervalArray
+
+        Returns
+        -------
+        IntervalArray
+        """
+        closed = set(interval.closed for interval in to_concat)
+        if len(closed) != 1:
+            raise ValueError("Intervals must all be closed on the same side.")
+        closed = closed.pop()
+
+        # TODO: avoid intermediate list
```

Comment:
This TODO (which I think I added) seems unavoidable, right?

Reply:
Yeah, I don't think it can be avoided. Removed the TODO.  IIUC the intermediate list doesn't copy data, so not a big deal (`np.concatenate` does of course).